File: web/src/app/modules/prets/pret.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { ApiResponse, PageResponse } from '@core/models/api-response.model';
import { PretResponse, StatutPret } from '@core/models/pret.model';

@Injectable({ providedIn: 'root' })
export class PretService {

  private readonly API = '/api/prets';

  constructor(private http: HttpClient) {}

  listPrets(statut?: StatutPret, page = 0, size = 20): Observable<PageResponse<PretResponse>> {
    let params = new HttpParams().set('page', page).set('size', size);
    if (statut) params = params.set('statut', statut);
    return this.http.get<ApiResponse<PageResponse<PretResponse>>>(this.API, { params }).pipe(
      map(r => r.data)
    );
  }

  getById(idPret: string): Observable<PretResponse> {
    return this.http.get<ApiResponse<PretResponse>>(`${this.API}/${idPret}`)
      .pipe(map(r => r.data));
  }

  getPretsClient(idClient: string): Observable<PretResponse[]> {
    return this.http.get<ApiResponse<PretResponse[]>>(`${this.API}/client/${idClient}`)
      .pipe(map(r => r.data));
  }
}

File: mobile/lib/core/services/pret_service.dart
import '../models/pret.dart';
import '../models/echeance.dart';
import '../models/page_response.dart';
import 'api_service.dart';

class PretService {
  final ApiService _api;

  PretService(this._api);

  Future<PageResponse<Pret>> getPrets({
    String? statut,
    int page = 0,
    int size = 20,
  }) async {
    final params = <String, dynamic>{
      'page': page,
      'size': size,
    };
    if (statut != null && statut.isNotEmpty) params['statut'] = statut;

    return _api.get<PageResponse<Pret>>(
      '/api/v1/prets',
      queryParameters: params,
      fromJson: (data) => PageResponse.fromJson(
        data as Map<String, dynamic>,
        (json) => Pret.fromJson(json),
      ),
    );
  }

  Future<Pret> getPretDetail(int idPret) async {
    return _api.get<Pret>(
      '/api/v1/prets/$idPret',
      fromJson: (data) => Pret.fromJson(data as Map<String, dynamic>),
    );
  }

  Future<List<Echeance>> getEcheances(int idPret) async {
    return _api.get<List<Echeance>>(
      '/api/v1/echeances/pret/$idPret',
      fromJson: (data) {
        final list = data as List<dynamic>;
        return list
            .map((e) => Echeance.fromJson(e as Map<String, dynamic>))
            .toList();
      },
    );
  }
}
